src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,395 @@
+  auto status = lease_client->RequestWorkerLease(
+      actor->GetCreationTaskSpecification(),
+      [this, node_id, retry_leasing_worker, actor](
+          const Status &status, const rpc::RequestWorkerLeaseReply &reply) {
+        if (auto node = gcs_node_manager_.GetNode(node_id)) {
+          if (!status.ok()) {
+            retry_leasing_worker();
+            return;
+          }
+          RAY_LOG(INFO) << "Finished leasing worker from " << node_id << " for actor "
+                        << actor->GetActorID();
+          HandleWorkerLeasedReply(actor, reply);
+        }
+      });
+
+  if (!status.ok()) {
+    retry_leasing_worker();
+  }
+}
+
+void GcsActorScheduler::HandleWorkerLeasedReply(
+    std::shared_ptr<GcsActor> actor, const ray::rpc::RequestWorkerLeaseReply &reply) {
+  // Remove the actor from the leasing queue.
+  RemoveActorInPhaseOfLeasing(actor);
+
+  const auto &retry_at_raylet_address = reply.retry_at_raylet_address();
+  const auto &worker_address = reply.worker_address();
+  if (worker_address.raylet_id().empty()) {
+    RAY_CHECK(!retry_at_raylet_address.raylet_id().empty());
+    auto address = actor->GetAddress();
+    address.set_raylet_id(retry_at_raylet_address.raylet_id());
+    actor->UpdateAddress(address);
+    auto actor_table_data =
+        std::make_shared<rpc::ActorTableData>(actor->GetActorTableData());
+    // The backend storage is reliable in the future, so the status must be ok.
+    RAY_CHECK_OK(actor_info_accessor_.AsyncUpdate(actor->GetActorID(), actor_table_data,
+                                                  [this, actor](Status status) {
+                                                    RAY_CHECK_OK(status);
+                                                    Schedule(actor);
+                                                  }));
+  } else {
+    std::vector<rpc::ResourceMapEntry> resources;
+    for (auto &resource : reply.resource_mapping()) {
+      resources.emplace_back(resource);
+    }
+    auto leased_worker = std::make_shared<GcsLeasedWorker>(
+        worker_address, std::move(resources), io_context_, client_factory_);
+    RAY_LOG(INFO) << "Worker " << leased_worker->GetWorkerID()
+                  << " is responsible for creating actor " << actor->GetActorID();
+    AddWorkerInPhaseOfCreating(leased_worker);
+    CreateActorOnWorker(actor, leased_worker);
+  }
+}
+
+void GcsActorScheduler::CreateActorOnWorker(std::shared_ptr<GcsActor> actor,
+                                            std::shared_ptr<GcsLeasedWorker> worker) {
+  RAY_CHECK(actor && worker);
+  actor->UpdateAddress(worker->GetAddress());
+  RAY_LOG(INFO) << "Start creating actor " << actor->GetActorID() << " on worker "
+                << worker->GetWorkerID() << " at node " << actor->GetNodeID();
+  worker->CreateActor(actor->GetCreationTaskSpecification(), [this, actor, worker] {
+    // This callback is invoked only when the actor is created successfully.
```

Comment:
I notice that this retries under the hood, but this is a bad idea because this could hang. For example, if the node or worker is removed while this request is pending, it looks like we will keep trying the request forever.

Please instead make the logic in GcsLeasedWorker simpler (don't retry) and propagate the status up to this callback. Then, check whether the actor is still in the node_to_workers_when_creating_ map before calling the success handler.

Also, please add a unit test for this.

Reply:
> I notice that this retries under the hood, but this is a bad idea because this could hang. For example, if the node or worker is removed while this request is pending, it looks like we will keep trying the request forever.
> 
> Please instead make the logic in GcsLeasedWorker simpler (don't retry) and propagate the status up to this callback. Then, check whether the actor is still in the node_to_workers_when_creating_ map before calling the success handler.
> 
> Also, please add a unit test for this.

I think there is no problem here.
When node or worker is remove the `assigned_actor_id` inside worker will be set to nil and the retry logic will exit when detect that the `assigned_actor_id` is nil.

Reply:
> I notice that this retries under the hood, but this is a bad idea because this could hang. For example, if the node or worker is removed while this request is pending, it looks like we will keep trying the request forever.
> 
> Please instead make the logic in GcsLeasedWorker simpler (don't retry) and propagate the status up to this callback. Then, check whether the actor is still in the node_to_workers_when_creating_ map before calling the success handler.
> 
> Also, please add a unit test for this.

Okey, I will fix it.